x-pack/plugins/security/server/deprecations/kibana_dashboard_only_role.ts

```diff
@@ -0,0 +1,244 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type {
+  SecurityGetRoleMappingResponse,
+  SecurityGetUserResponse,
+} from '@elastic/elasticsearch/api/types';
+
+import { i18n } from '@kbn/i18n';
+import type {
+  DeprecationsDetails,
+  DeprecationsServiceSetup,
+  ElasticsearchClient,
+  Logger,
+  PackageInfo,
+} from 'src/core/server';
+
+import type { SecurityLicense } from '../../common';
+import { getDetailedErrorMessage, getErrorStatusCode } from '../errors';
+
+export const KIBANA_DASHBOARD_ONLY_USER_ROLE_NAME = 'kibana_dashboard_only_user';
+
+export interface Deps {
+  deprecationsService: DeprecationsServiceSetup;
+  license: SecurityLicense;
+  logger: Logger;
+  packageInfo: PackageInfo;
+}
+
+export const registerKibanaDashboardOnlyRoleDeprecation = ({
+  deprecationsService,
+  logger,
+  license,
+  packageInfo,
+}: Deps) => {
+  deprecationsService.registerDeprecations({
+    getDeprecations: async (context) => {
+      // Nothing to do if security is disabled
+      if (!license.isEnabled()) {
+        return [];
+      }
+
+      return [
+        ...(await getUsersDeprecations(context.esClient.asCurrentUser, logger, packageInfo)),
+        ...(await getRoleMappingsDeprecations(context.esClient.asCurrentUser, logger, packageInfo)),
+      ];
+    },
+  });
+};
+
+async function getUsersDeprecations(
+  client: ElasticsearchClient,
+  logger: Logger,
+  packageInfo: PackageInfo
+): Promise<DeprecationsDetails[]> {
+  let users: SecurityGetUserResponse;
+  try {
+    users = (await client.security.getUser()).body;
+  } catch (err) {
+    logger.error(`Failed to retrieve users: ${getDetailedErrorMessage(err)}`);
+    return deprecationError(packageInfo, err);
+  }
+
+  const usersWithKibanaDashboardOnlyRole = Object.values(users)
+    .filter((user) => user.roles.includes(KIBANA_DASHBOARD_ONLY_USER_ROLE_NAME))
+    .map((user) => user.username);
+  if (usersWithKibanaDashboardOnlyRole.length === 0) {
+    return [];
+  }
+
+  return [
+    {
+      title: i18n.translate(
+        'xpack.security.deprecations.kibanaDashboardOnlyUser.deprecationTitle',
+        {
+          defaultMessage: 'The "{roleName}" role is removed',
+          values: { roleName: KIBANA_DASHBOARD_ONLY_USER_ROLE_NAME },
+        }
+      ),
+      message: i18n.translate(
+        'xpack.security.deprecations.kibanaDashboardOnlyUser.usersDeprecationMessage',
+        {
+          defaultMessage:
+            'The following users have a deprecated and removed "{roleName}" role: {users}. Create a custom role with Kibana privileges to restrict access to just the Dashboard feature instead.',
+          values: {
+            roleName: KIBANA_DASHBOARD_ONLY_USER_ROLE_NAME,
+            users: usersWithKibanaDashboardOnlyRole.join(', '),
+          },
+        }
+      ),
+      level: 'warning',
```

Comment:
__question:__ if I understand correctly we should not use `critical` here since it doesn't really blocks upgrade?